### ISSUE TITLE
Refinements for field selectors

### ISSUE THREAD

**Issue by jprider63** [ucsd-progsys/liquidhaskell#1384]
We discussed this yesterday, but I'm opening an issue to elaborate and so we don't forget. We want LH to automatically generate refinements for field selectors. In addition, users currently can specify their own refinements that are unchecked, which is [unsound](https://github.com/ucsd-progsys/liquidhaskell/blob/2820b7530ddd09e289c60df4ce085faa7c507231/tests/refined-classes/semigroup_unsound.hs#L101). It looks like #595 is related, but it seems this wasn't fully addressed. 

Here are a few examples of how I think the generated refinements should look like. For `D1`:

```
{-@
data D1 a =
    S1 {d1Fst :: {v:a | x a}}
  | D1 {d1Fst :: {v:a | y a}, d1Snd :: {v:a | z a}}
  | N
@-}
data D1 a =
    S1 {d1Fst :: a}
  | D1 {d1Fst :: a, d1Snd :: a}
  | N

{-@ d1Fst
 :: d:{D1 a | isS1 d || isD1 d}
 -> r:{a | (isS1 d => x a) && (isD1 d => y a)}
 @-}

{-@ d1Snd
 :: d:{D1 a | isD1 d}
 -> r:{a | z a}
 @-}
```

There are preconditions on the `D1 a` arguments based on which constructors are valid. There is also an implication from which constructor is given to the refinement output for that constructor. When there is only one valid constructor (ex, `d1Snd`), the implication in the resulting refinement can be omitted. 

Another example:

```
{-@
data D2 a =
    D2 {d2Fst :: {v:a | y a}}
@-}
data D2 a =
    D2 {d2Fst :: a}

{-@ d2Fst
 :: D2 a
 -> {v:a | y a}
 @-}
```

In `d2Fst`, all of the constructors of `D2` are valid, so you can omit the constructor precondition.

**Comment by ranjitjhala** [comment]
@nikivazou should this be closed by your recent PR #1385

**Comment by nikivazou** [comment]
No,that pr was only for the unsound example... Still working on that
